modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java

```diff
@@ -152,46 +207,56 @@ SynonymMap buildSynonyms(Analyzer analyzer, ReaderWithOrigin rules) {
         }
     }
 
-    protected ReaderWithOrigin getRulesFromSettings(Environment env, IndexCreationContext context) {
-        if (settings.getAsList("synonyms", null) != null) {
-            List<String> rulesList = Analysis.getWordList(env, settings, "synonyms");
-            StringBuilder sb = new StringBuilder();
-            for (String line : rulesList) {
-                sb.append(line).append(System.lineSeparator());
-            }
-            return new ReaderWithOrigin(new StringReader(sb.toString()), "'" + name() + "' analyzer settings");
-        } else if (settings.get("synonyms_set") != null) {
-            if (analysisMode != AnalysisMode.SEARCH_TIME) {
-                throw new IllegalArgumentException(
-                    "Can't apply [synonyms_set]! " + "Loading synonyms from index is supported only for search time synonyms!"
-                );
-            }
-            String synonymsSet = settings.get("synonyms_set", null);
-            // provide fake synonyms on index creation and index metadata checks to ensure that we
-            // don't block a master thread
-            if (context != IndexCreationContext.RELOAD_ANALYZERS) {
-                return new ReaderWithOrigin(
-                    new StringReader("fake rule => fake"),
-                    "fake [" + synonymsSet + "] synonyms_set in .synonyms index",
-                    synonymsSet
-                );
-            }
-            return new ReaderWithOrigin(
-                Analysis.getReaderFromIndex(synonymsSet, synonymsManagementAPIService),
-                "[" + synonymsSet + "] synonyms_set in .synonyms index",
+    record ReaderWithOrigin(Reader reader, String origin, String resource) {
+        ReaderWithOrigin(Reader reader, String origin) {
+            this(reader, origin, null);
+        }
+    }
+
+    private static ReaderWithOrigin getRulesReaderInline(SynonymTokenFilterFactory factory, IndexCreationContext context) {
```

Comment:
These methods could be inlined directly in the enum instances to make it explicit there's a change in behaviour in the different instances, see https://github.com/Mikep86/elasticsearch/pull/2. That way we get rid of the provider in the constructor which is clearer IMO